docs/design/milestone_2.md

```diff
@@ -0,0 +1,260 @@
+
+The GitHub enumeration tool will be altered in the following way:
+
+- Support environment variables along with flags for easy configuration through
+  kubeconfig YAML files.
+- The ability to specify a destination bucket and file for project URLs. `runID`
+  can be specified using a token (e.g. `[[run-id]]`).
+- The ability to specify a destination bucket and file for the marker file.
+
+Cloud infrastructure changes:
+
+- A service worker needs to be associated with the enumeration.
+- A bucket with write access by the above service worker.
+  - A lifecycle policy on the bucket to clean up stale data.
+- A Kubernetes configuration with the following:
+  - The desired schedule
+  - The destination bucket and path structure
+  - The address + port of the Scorecard GitHub token auth server
+
+#### Out of Scope for Enumeration
+
+##### Additional Reliability and Performance
+
+GitHub enumeration is reasonably reliable already, however it is still
+susceptible to failures (e.g. network, cloud storage).
+
+Given that the GitHub enumeration only takes 4-6 hours to complete for all
+projects with at least 10 stars, additional efforts to improve reliability will
+be deferred.
+
+In the future an architecture similar to the Signal Collection and Scorecard
+project can be used to improve performance and reliability.
+
+##### Deduplication
+
+Deduplication of URLs from different sources is out of scope. Long term this
+will be important to ensure API quotas are used efficiently. For Milestone 2,
+deduplication will be done when results are being reduced as this needs to be
+handled anyway (e.g. a timeout triggers a second set of signals to be collected
+for the same repository).
+
+### Signal Collection
+
+To scale signal collection we will build upon the Scorecard project's
+infrastructure.
+
+The
+[Scalable Scorecard](https://github.com/ossf/scorecard/blob/main/docs/design/scalable_scorecards.md)
+covers the specifics of the Scorecard infrastructure.
+
+To use the Scorecard infrastructure we will need to make the following changes
+to the Scorecard project:
+
+- Controller URL iteration ([ref](https://github.com/ossf/scorecard/blob/main/cron/internal/data/iterator.go))
+  - Identify the prefix to search from the marker file.
+  - Read from `bucket/{runID}/*.csv` instead of the static CSV file.
+- Worker library ([ref](https://github.com/ossf/scorecard/blob/main/cron/internal/worker/main.go))
+  - Create a reusable library for:
+    - consuming PubSub messages containing work shards
+    - writing the shards to storage
+- Data-transfer library ([ref](https://github.com/ossf/scorecard/blob/main/cron/internal/bq/main.go))
+  - A reusable library for determining if all shards are present, and: iterating
+    over the files (e.g. BigQuery loader) and/or iterating over each record in
+    the resulting CSV files.
+
+Additionally, the following changes need to be made to the Criticality Score
+project:
+
+- Implement a criticality score specific worker
```

Comment:
nit: link to the final implementation of the worker?

Reply:
Done